Reduxify setting the post geolocation

jsnajdr opened the pull request:
Quite a big PR, the main one in the "Post Metadata R12n" series. Several commits:

**Add Redux logic for editing post metadata**

Add two new actions creators, `updatePostMetadata` and `deletePostMetadata` that produce `POST_EDIT` actions with the appropriate properties. Equivalent to the `updateMetadata` and `deleteMetadata` Flux actions.

The `state.posts.edits` reducer now correctly merges the incoming post updates into the metadata edits, leaving only the unapplied ones.

The `isEditedPostDirty` selector compares the outstanding edits to the original post metadata values and determines if the edits are noops or not, affecting the return values of the selector.

The `getEditedPost` selector merges the original saved post object with the outstanding edits and returns a post object will most recent metadata.

All new Redux logic is covered by unit tests.

**Use Redux to store copied metadata when starting to edit a post copy**

When starting to edit a post copy, we copy some attributes to the Flux store and some to the Redux store, depending on how exactly editing of that particular attribute is implemented at the moment.

This patch moves the copied metadata to Redux. The only metadata we copy right now are the geolocation fields, and these are being moved to Redux in the next patch.

**Reduxify setting the post geolocation**

Use the new Redux metadata logic to edit post geolocation: get data from Redux selectors and do edits with Redux action dispatches.

**Send geolocation updates as formatted strings to ensure correct comparison**

Fixes a bug I found during testing. The float values of geo coordinates should be formatted as strings with 7 decimal digits before sending them as metadata values to REST API. This ensures correct comparison between the values in the edits and the values returned by REST API responses. The responses use the string format described above.

Without this patch, geolocation updates are never marked as "saved", because the values are different.

**How to test:**
On both new and existing posts, test the following:

1. Can you set the geolocation of a post, change it and unset it? Do the geolocation data get correctly saved? After changing the geolocation, does the editor show the post as "dirty"? After save or autosave, is the "dirty" flag reset? Watch for the "Save" button in the masterbar for drafts. For published posts, try to close the editor without saving. Do and don't you get a confirmation dialog for "dirty" and "non-dirty" posts?

2. Can you still edit metadata-powered fields that are not Redux yet? These are "SEO Description" and "Sharing" sections (Publicize). Do these sections still work? Do "SEO Description", "Sharing" and "Location" edits work together well?

3. When editing post copy as a new post, does the geolocation info get properly copied over?

jsnajdr committed:
Add Redux logic for editing post metadata

Add two new actions creators, `updatePostMetadata` and `deletePostMetadata` that
produce `POST_EDIT` actions with the appropriate properties. Equivalent to the
`updateMetadata` and `deleteMetadata` Flux actions.

The `state.posts.edits` reducer now correctly merges the incoming post updates
into the metadata edits, leaving only the unapplied ones.

The `isEditedPostDirty` selector compares the outstanding edits to the original
post metadata values and determines if the edits are noops or not, affecting the
return values of the selector.

All new Redux logic is covered by unit tests.

jsnajdr committed:
Use Redux to store copied metadata when starting to edit a post copy

When starting to edit a post copy, we copy some attributes to the Flux store
and some to the Redux store, depending on how exactly editing of that particular
attribute is implemented at the moment.

This patch moves the copied metadata to Redux. The only metadata we copy right
now are the geolocation fields, and these are being moved to Redux in the next patch.

jsnajdr committed:
Reduxify setting the post geolocation

Use the new Redux metadata logic to edit post geolocation: get data from Redux
selectors and do edits with Redux action dispatches.

jsnajdr committed:
Send geolocation updates as formatted strings to ensure correct comparison

Formats float values of geo coordinates to strings with 7 decimal digits before
sending them as metadata values to REST API. This ensures correct comparison between
the values in the edits and the values returned by REST API responses. The responses
use the string format described above.

Without this patch, geolocation updates are never marked as "saved", because the
values are different.

blowery commented:
When copying a post, I see the location properly in the UI, but after publishing and reloading the post, the location is no longer set. 

http://calypso.localhost:3000/post/b19ytest.wordpress.com/289

blowery commented:
Don't think it's this PR, but the autocomplete on the location search is really wonky. Trying to find `Manchester, MI` has proven quite difficult. There's only ever one suggestion and it's rarely correct.

blowery commented:
One more bug:
* make post with a location 
* make another post, copy from the first post
* publish

On trying to close the preview window, I get the "unsaved changes" confirmation dialog. Moving into the edit page, clicking update, wait, click close, still get the confirmation dialog. The post seems to be eternally dirty.

jsnajdr committed:
When copying post metadata, ignore the 'id' attribute

Metadata edits that have an 'id' attribute are ignored by the server and
the edit never gets saved.

jsnajdr commented:
Nice catch @blowery with the two bugs in post copying. They are actually one bug, caused by the fact that I copy all metadata object properties from the original post, including `id`. The server then ignores these updates and they are never saved, keeping the post eternally dirty. Fixed by including only the `key` and `value` attributes. The original Flux code behaved the same way, although it's not immediately obvious.

> Don't think it's this PR, but the autocomplete on the location search is really wonky. Trying to find Manchester, MI has proven quite difficult. There's only ever one suggestion and it's rarely correct.

We use Google Maps API for suggestions based on geo coordinates and it indeed returns only one suggestion. In your case, unfortunately it's Michigan, UK 😢 

Ready for more review.

nylen commented:
At first glance this PR looks a bit daunting, but over half of the new code is unit tests for core application functionality:

```sh
export MERGE_BASE=$(git merge-base origin/master HEAD)
git diff --stat $MERGE_BASE $(git diff --name-only $MERGE_BASE | grep -v /test/)
git diff --stat $MERGE_BASE $(git diff --name-only $MERGE_BASE | grep /test/)

# Application code changes :  7 files changed, 223 insertions(+), 40 deletions(-)
# Test code changes        :  4 files changed, 339 insertions(+), 1 deletion(-)
```

> After changing the geolocation, does the editor show the post as "dirty"?

I think this PR is a good example of how poorly-defined behavior in our application code leads to bugs like the "Are you sure?" editor prompt (#9833), and how to find and fix specific instances of this bug.  Also, the new unit tests should ensure that this particular bug will not reoccur later.

---

From a pretty quick read through the code, one thing I see that is a bit confusing is the concept of "applied" vs "unapplied" post metadata edits.

For example, in the unit test for `should return false if all metadata edits are already applied`, the Redux state indicates that the geo-location meta has been deleted on the client, but not saved (it's in `edits`), and this deletion has already taken place on the server.  In what circumstances would this occur?

---

Otherwise, this looks good from a quick read, and I will also spend some time today testing it out.

jsnajdr commented:
> For example, in the unit test for should return false if all metadata edits are already applied, the Redux state indicates that the geo-location meta has been deleted on the client, but not saved (it's in edits), and this deletion has already taken place on the server. In what circumstances would this occur?

This can typically occur when the user changes a value and then changes it back to the original value. Then there is a noop `update` operation in the `edits`. Or the user can add a meta and then delete it. That's how geo-location is edited, by adding and removing the meta properties. `edits` contain a noop `delete` operation in that case.

In both cases, we want to show the post as "not dirty".

Why do we keep a noop update/delete operation instead of removing it from `edits` completely? Because when the edit is requested by the user, there might be save/autosave happening in the background, and the REST API request returns the full post object as a response value. Then we need to correctly apply the outstanding edits to the new "saved post" object.

The original Flux store achieved that by queueing the edit actions while a save is in progress, and then applying (similarly to Git rebase) them onto the new "saved post".

Redux implementation doesn't do any queueing, but keeps the `edits` as a separate object and the reducer removes noop edits when receiving the new "saved post". Slightly different, but achieves the same outcome.

1. Start saving a post that has `title = "hello"`
2. Perform an edit setting `title` to `hello`, i.e., a noop one
3. Save finished, received post with `title = "hello world"`. Maybe someone else changed the title in meantime, or the attribute is some tricky one that gets autoupdated by the server.

What should we do with the outstanding `title ="hello"` edit? Should we (A) forget it because it was noop, or (B) to perform it and change `"hello world"` back to `"hello"`?

Neither choice is good or bad per se, but the Calypso behavior has always been (B) and the Redux work doesn't have an ambition to change it.

blowery commented:
LGTM!
